Refactor makesyscalls.lua

agge3 opened the pull request:
Refactor of makesyscalls.lua to be a library
Drop-in replacement to makesyscalls.lua

Google Summer of Code 2024 Final Work Product
Project description: [Refactor Syscall Creation Script](https://wiki.freebsd.org/SummerOfCode2024Projects/RefactorSyscallCreationScript)
Mentors: Warner Losh (bsdimp), Kyle Evans (kevans)

Sponsored by:    Google (GSoC 2024)

agge3 commented:
Yes. Thank you Warner! I'm excited to get this where it needs to be. I'll appreciate any time and effort spent.

There are bits and pieces that are still being iterated on, but the guts are there. Files are being generated (mostly) similar.

-Tyler

brooksdavis reviewed (commented):
This is quite hard to review in any kind of meaningful way.  There needs to be some sort of overview of the structure in a README.md as well as the commit message.

What I'll want to do before a merge is to take something closer to final and bring it over to CheriBSD so I can apply our local diffs and see how that works in practice.

Some comments:
 - Despite all the indentation commits (that need to be squashed), indentation is wildly inconsistent even within files.
 - This is missing some commits to makesyscalls.lua (FREEBSD14 is the canary, but everything since the original split needs a review and merge).
 - If there are any differences in generated output they need to be very well documented. The desired number is zero.
 - The code might benefit from some subdirectories to differentiate code that generates specific files from library code.
 - There should probably be three total commits: add new code, switch what the Makefile uses, remove old code.

kevans91 commented on sys/tools/syscalls/bsdio.lua:
```diff
@@ -0,0 +1,144 @@
+--
+-- SPDX-License-Identifier: BSD-2-Clause
+--
+-- Copyright (c) 2024 Tyler Baxter <agge@FreeBSD.org>
+--
+--
+-- Thanks to Kyle Evans for his makesyscall.lua in FreeBSD which served as
+-- inspiration for this, and as a source of code at times.
+--
+-- SPDX-License-Identifier: BSD-2-Clause-FreeBSD
+--
+-- Copyright (c) 2019 Kyle Evans <kevans@FreeBSD.org>
+--
```
Yeah, go for it

agge3 replied:
Great! This has been done. I've kept the history bit in the main file.

kevans91 commented on sys/tools/syscalls/bsdio.lua:
```diff
@@ -0,0 +1,144 @@
+--
+function bsdio:pad64(bool)
+    if bool then
+	    self:write(string.format([[
```
I don't think you're getting anything out of `string.format()` here -- let's drop that and just use the multiline string directly as an argument to write()

agge3 replied:
Yes, I found other instances of this. Moving forward, please let me know if you see anything else.

kevans91 commented on sys/tools/syscalls/bsdio.lua:
```diff
@@ -0,0 +1,144 @@
+-- in.
+-- Useful when a file has multiple stages of generation.
+function bsdio:store(str, level)
+    level = level or 1 -- default to level one if not provided
```
Let's just consolidate the comment on this line up with that above the function. Maybe:

```
Can also specify which order (level) to store in, which defaults to one if not provided.
```

agge3 replied:
There are other comments that could use a similar suggestion. I've changed those to match this suggestion (and this one). Please inform me if you see anything else.

kevans91 commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,321 @@
+		return {}
+	end
+
+	local fh = io.open(file)
```
`io.open` returns nil, msg, errno in the case of failure.  We should at least preserve the msg so that we can pass it on in the `abort()` below in the `not open_fail_ok` case

kevans91 commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,321 @@
+		end
+	end
+
+	assert(io.close(fh))
```
```suggestion
	assert(fh:close())
```

We were admittedly inconsistent about this in makesyscalls.lua, but we should try to converge on this.

agge3 replied:
Following up on these suggestions, since it's been a couple weeks. Pretty much, all of these are will be done. After discussing with Warner, my main focus is on reducing the diffs. Other suggestions I'm paying attention to though, and I'll be incorporating them as I'm working in that area.

agge3 replied:
This is consistent now.

kevans91 commented on sys/tools/syscalls/freebsd-syscall.lua:
```diff
@@ -0,0 +1,134 @@
+
+	self.syscalls = { }
+
+	local fh = io.open(file)
```
Ditto note about preserving the failure msg for the failure note below.  Even better here might be converting all of these print + return {} combo to return `nil, <msg>` on failure to avoid the library printing things that the frontend doesn't want and give the consumer control + ability to use the standard assert() pattern.

agge3 replied:
Yes! These have been changed. The library should not be printing anything. The other suggestions in this thread have also been incorporated. Thank you Kyle!

kevans91 reviewed (commented):
Initial pass with some feedback; apologies for largely being a ghost this term.  I'll try to scrape some time together in the next week or so to do a more thorough review.

jlduran commented:
It would be good to first address [luacheck](https://github.com/mpeterv/luacheck)'s suggestions:
```
luacheck sys/tools/syscalls
```

bsdimp commented:
I've noticed 2 real differences so far (the other differences are comments or sorting deltas):

We have `#define nosys linux_nosys` added to the linux files. what is this done? It doesn't seem to be needed.
The return type of linux_exit changes from int to void, which is incorrect.

There may be other issues as well.

agge3 commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,321 @@
+		{ stdcompat = "FREEBSD10" },
+		{ stdcompat = "FREEBSD11" },
+		{ stdcompat = "FREEBSD12" },
+		{ stdcompat = "FREEBSD13" },
```
Included. Thank you.

agge3 commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,321 @@
+-- merged into the global config.
+-- Helper function for config.capability() to use.
+--
+-- NOTE: Hasn't been changed from makesyscalls.lua, will work the same.
```
There were other porting notes. I've removed them.

agge3 commented on sys/tools/syscalls/bsdio.lua:
```diff
@@ -0,0 +1,144 @@
+
+util = require("util")
+
+local bsdio = {}
```
I'm planning on renaming to `generator`, since much of its use is binding to a file and writing. I'm also considering splitting off the macros into macros.lua (or re-evaluating if they provide value). Let me know if this sounds better.

brooksdavis commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,288 @@
+    ptr_intptr_t_cast = "intptr_t",
+    obsol = {},
+    unimpl = {},
+    capabilities_conf = "capabilities.conf",
```
Support for capabilities.conf should be dropped as the last consumer went away with be67ea40c5a0fd375d754a92e6cd8a3bfc508ba3. It should provably not be removed as part of the refactor, but done on one side or the other.

brooksdavis commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,288 @@
+
+-- Instantiates config.compat_options.
+function config.mergeCompat()
+    if config.compat_set ~= "" then
+        if not compat_option_sets[config.compat_set] then
+            util.abort(1, "Undefined compat set: " .. config.compat_set)
+        end
+
+        config.compat_options = compat_option_sets[config.compat_set]
+    end
```
spaces should be tabs

brooksdavis commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,288 @@
+
+-- Merge capability (Capsicum) configuration into the global config.
+function config.mergeCapability()
+    -- We ignore errors here if we're relying on the default configuration.
+    if not config.modifications.capenabled then
+        config.capenabled = grabCapenabled(config.capabilities_conf,
+            config.modifications.capabilities_conf == nil)
+    elseif config.capenabled ~= "" then
+        -- We have a comma separated list from the format of capabilities.conf,
+        -- split it into a set with boolean values for each key.
+        config.capenabled = util.setFromString(config.capenabled, "[^,]+")
+    end
```
spaces that should be tabs

brooksdavis commented on sys/tools/syscalls/core/freebsd-syscall.lua:
```diff
@@ -0,0 +1,126 @@
+		-- NOTE: Can't use pure pattern matching here because of the 's' test
+		-- and this is shorter than a generic pattern matching pattern
+		if line == nil or line == "" then
+            goto skip -- Blank line, skip this line.
```
spaces that should be tabs

brooksdavis commented on sys/tools/syscalls/core/freebsd-syscall.lua:
```diff
@@ -0,0 +1,126 @@
+			if s:add(line) then
+				-- Append to system call list.
+				for t in s:iter() do
+                    if t:validate(t.num - 1) then
+					    table.insert(self.syscalls, t)
+                    else
+                        util.abort(1, "Skipped system call at number " .. t.num)
+                    end
```
more spaces. There are several more instances in this file and presumably others, but I'm going to stop commenting here.

agge3 replied:
In regards to spaces that should be tabs. I've changed all of them to tabs--but programmatically. If I catch any (which I haven't) I will continue to change. This should be done.

brooksdavis commented on sys/tools/syscalls/core/syscall.lua:
```diff
@@ -0,0 +1,444 @@
+
+-- CREDIT: http://lua-users.org/wiki/CopyTable
+-- Save copied tables in `copies`, indexed by original table.
+function deepCopy(orig, copies)
```
It doesn't seem right to have two slightly different implementations of deepCopy()

brooksdavis commented on sys/tools/syscalls/core/syscall.lua:
```diff
@@ -0,0 +1,444 @@
+--
+-- As we're parsing the system calls, there's two types. Either we have a
+-- specific one, that's a assigned a number, or we have a range for things like
+-- reseved system calls. this function deals with both knowing that the specific
+-- ones are more copy and so we should just return the object we just made w/o
+-- an extra clone.
```
I'm not sure what this sentence is trying to say.

agge3 replied:
I've reworded. Please let me know if it is still unclear.

brooksdavis commented on sys/tools/syscalls/scripts/init_sysent.lua:
```diff
@@ -0,0 +1,221 @@
+
+-- xxx need compat call count
+
+-- Should be the same as makesyscalls.lua generates, except that we don't bother
+-- to align the system call stuff... it's badly broken anyway and looks like crap
+-- so we're declaring that a bug and removing all that crazy book-keeping to.
+-- If we need to do it, and I hope we don't, I'll just create a string and do
+-- #str to figure out how many tabs to add
```
I agree the current formatting is low value. If it's too ugly to make the output the same here, consider removing the alignment from makesyscalls.lua before the switch. The code change should be reviewable there whereas diffs to the generated files are impossible to review.

agge3 replied:
I like this suggestion a lot. When this PR is closer to where it needs to be, I want to revisit this and open a PR to remove the aligning from makesyscalls.lua. I would like Kyle's input on this.

brooksdavis commented on sys/tools/syscalls/main.lua:
```diff
@@ -0,0 +1,63 @@
+-- inspiration for this, and as a source of code at times.
+--
+
+-- We generally assume that this script will be run by flua, however we've
+-- carefully crafted modules for it that mimic interfaces provided by modules
+-- available in ports.  Currently, this script is compatible with lua from ports
+-- along with the compatible luafilesystem and lua-posix modules.
+
```
I'd remove copies of this syscalls/scripts/*.lua. I might also move this copy to the README.md.

brooksdavis reviewed (commented):
I like the code reorganization. Much improved!

My main gripe: pick an indent style and use it consistently. I personally prefer a tabbed style following our C practices, but 4-spaces would be fine.

agge3 commented:
> It would be good to first address [luacheck](https://github.com/mpeterv/luacheck)'s suggestions:
> 
> ```
> luacheck sys/tools/syscalls
> ```

This has been followed up on. All luacheck suggestions are addressed. Thanks for the suggestion! This will be the precedent moving forward.

agge3 commented:
> I like the code reorganization. Much improved!
> 
> My main gripe: pick an indent style and use it consistently. I personally prefer a tabbed style following our C practices, but 4-spaces would be fine.

Great!! I agree with this sentiment. I will be following your preference. I did not catch that (also the source of why there's differences, as code from makesyscalls.lua followed FreeBSD C practices). It's going to be something I do as I correct the diffs, so it probably won't be changed right away, but it should follow a consistent indentation when it's ready.

agge3 commented:
I'd like to request more review.

I've opened a separate pull request, [sysent: Remove comment aligning #1441](https://github.com/freebsd/freebsd-src/pull/1441), to remove the comment aligning from sysent. The generated diffs should be compared to that.

The diffs have been vastly reduced. Currently (to my eye), there is one code diff: in sysent for native/freebsd32, there's a preprocessor macro for FREEBSD14 being added. As FREEBSD14 is defined in compat_options, I'm not sure if that is intentional or not, so I'd like some clarification on that.

The other diffs are exclusively comments, which I will be working on more (names not filtering correctly).

I have worked to incorporate all prior review suggestions. I agree with Brooks that changing makesyscalls.lua to syscalls/main.lua in the makefile should be a separate commit. I am leaving it in for the time being for ease of generating. That will be a final step.

I feel the library is much closer to where it needs to be, so I'd like some review on "structure, lua goodness, etc" and the current generated diffs.

There was a certain amount of "getting up to speed quickly", so please inform me of any suggestions I need to implement. I'm excited to contribute meaningful and useful work to the FreeBSD project.

Thank you! I appreciate your time!

brooksdavis reviewed (commented):
This is generally looking good. With changes like those I've proposed and a just committed upstream patch to reduce preamble diffs, the level of churn in the output is acceptable (and entirely a no-op.)

I'd ideally like to try integrating this into CheriBSD before a merge, but that shouldn't be a blocker if I don't get to it in the next week.

brooksdavis commented:
Just noticed while looking to apply CheriBSD changes: This is missing the functionality added in b71defdbd8715c1a778ebf8195e41b1c6db90d6c (with follow on 304cdac7385a40477484737fc0a4aa928c3be9d6) and 79b758e327757dc015fb26a122fdc9608c405e9e. I may try to add it tomorrow.

agge3 commented:
> Just noticed while looking to apply CheriBSD changes: This is missing the functionality added in [b71defd](https://github.com/freebsd/freebsd-src/commit/b71defdbd8715c1a778ebf8195e41b1c6db90d6c) (with follow on [304cdac](https://github.com/freebsd/freebsd-src/commit/304cdac7385a40477484737fc0a4aa928c3be9d6)) and [79b758e](https://github.com/freebsd/freebsd-src/commit/79b758e327757dc015fb26a122fdc9608c405e9e). I may try to add it tomorrow.

OK. Unfortunately Warner and Kyle were very busy this GSoC, so there was some degree of unanswered questions (although they tried their best). This is a problem that I can address.

Let me know the results of tomorrow, otherwise I will plan to incorporate.

brooksdavis commented on sys/tools/syscalls/core/scarg.lua:
```diff
@@ -0,0 +1,153 @@
+--
+-- SPDX-License-Identifier: BSD-2-Clause
+--
+-- Copyright (c) 2024 Tyler Baxter <agge@FreeBSD.org>
+-- Copyright (c) 2023 Warner Losh <imp@bsdimp.com>
+-- Copyright (c) 2019 Kyle Evans <kevans@FreeBSD.org>
```
To pick a file where the original code is mostly mine or derived from mine, there should likely be an SRI copyright line in the general copyright blocks:
```
-- Copyright (c) 2021-2024 SRI International
```

agge3 replied:
I've added this copyright to scarg.lua and config.lua. There's a reasonable amount of your work in config.lua. Let me know any other locations!

brooksdavis commented:
> I've added the missing support in https://github.com/brooksdavis/freebsd/tree/refactor-makesyscalls.lua

Just noticed I'd failed to commit the core implementation of the _libsys_h generation so you'll want to make sure you're at 1e1adfc1c4fe or later.

brooksdavis commented:
Wishlist item:
I accidentally wrote this declaration in CheriBSD and the old parser accepted it:
```
91      AUE_NULL        STD|CAPENABLED {
                int cheri_cidcap_alloc(
                    uintcap_t *cidp
                ) 
        }
```
The new one blew up messily:
```
/usr/libexec/flua: ../tools/syscalls/scripts/../core/scarg.lua:50: invalid pattern capture
stack traceback:
        [C]: in function 'string.gsub'
        ../tools/syscalls/scripts/../core/scarg.lua:50: in function 'core.scarg.init'
        ../tools/syscalls/scripts/../core/scarg.lua:148: in function 'core.scarg.new'
        ../tools/syscalls/scripts/../core/syscall.lua:259: in function 'core.syscall.addArgs'
        ../tools/syscalls/scripts/../core/syscall.lua:362: in function 'core.syscall.add'
        ../tools/syscalls/scripts/../core/freebsd-syscall.lua:62: in function 'core.freebsd-syscall.parseSysfile'
        ../tools/syscalls/scripts/../core/freebsd-syscall.lua:134: in function 'core.freebsd-syscall.new'
        ../tools/syscalls/scripts/syscalls.lua:90: in main chunk
        [C]: in ?
```
Catching this error might be with doing.

brooksdavis commented on sys/tools/syscalls/tools/util.lua:
```diff
@@ -0,0 +1,165 @@
+-- PARAM: abi, nil or optional ABI-specified intptr_t.
+--
+function util.isPtrType(type, abi)
+	local default = "intptr_t" or abi
```
```suggestion
	local default = abi or "intptr_t"
```
Currently default is always set to `"intptr_t"`

jlduran commented:
> Wishlist item: I accidentally wrote this declaration in CheriBSD and the old parser accepted it:
> 
> ```
> 91      AUE_NULL        STD|CAPENABLED {
>                 int cheri_cidcap_alloc(
>                     uintcap_t *cidp
>                 ) 
>         }
> ```
> 
> The new one blew up messily:
> 
> ```
> /usr/libexec/flua: ../tools/syscalls/scripts/../core/scarg.lua:50: invalid pattern capture
> stack traceback:
>         [C]: in function 'string.gsub'
>         ../tools/syscalls/scripts/../core/scarg.lua:50: in function 'core.scarg.init'
>         ../tools/syscalls/scripts/../core/scarg.lua:148: in function 'core.scarg.new'
>         ../tools/syscalls/scripts/../core/syscall.lua:259: in function 'core.syscall.addArgs'
>         ../tools/syscalls/scripts/../core/syscall.lua:362: in function 'core.syscall.add'
>         ../tools/syscalls/scripts/../core/freebsd-syscall.lua:62: in function 'core.freebsd-syscall.parseSysfile'
>         ../tools/syscalls/scripts/../core/freebsd-syscall.lua:134: in function 'core.freebsd-syscall.new'
>         ../tools/syscalls/scripts/syscalls.lua:90: in main chunk
>         [C]: in ?
> ```
> 
> Catching this error might be with doing.

I think it is because `)` is unescaped in the `gsub`:

```lua
> scarg = ")"
> scarg:gsub(scarg:match("([^* ]+)$") .. "$", "")
stdin:1: invalid pattern capture
stack traceback:
	[C]: in function 'string.gsub'
	stdin:1: in main chunk
	[C]: in ?
```

One simple fix is to escape it in double brackets:

```lua
> scarg:gsub([[scarg:match("([^* ]+)$")]] .. "$", "")
)	0
```

brooksdavis commented:
> > I've added the missing support in https://github.com/brooksdavis/freebsd/tree/refactor-makesyscalls.lua
> 
> Just noticed I'd failed to commit the core implementation of the _libsys_h generation so you'll want to make sure you're at 1e1adfc1c4fe or later.

I've found and fixed a few more issues while porting the various additions to makesyscalls.lua in CheriBSD:
 - util.isPtrType's optional arg handling was wrong
 - util.isPtrType needs its optional arg almost all the time. It's suboptimal that it's optional.
 - config.mincompat was ignored

agge3 commented:
> > Wishlist item: I accidentally wrote this declaration in CheriBSD and the old parser accepted it:
> > ```
> > 91      AUE_NULL        STD|CAPENABLED {
> >                 int cheri_cidcap_alloc(
> >                     uintcap_t *cidp
> >                 ) 
> >         }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The new one blew up messily:
> > ```
> > /usr/libexec/flua: ../tools/syscalls/scripts/../core/scarg.lua:50: invalid pattern capture
> > stack traceback:
> >         [C]: in function 'string.gsub'
> >         ../tools/syscalls/scripts/../core/scarg.lua:50: in function 'core.scarg.init'
> >         ../tools/syscalls/scripts/../core/scarg.lua:148: in function 'core.scarg.new'
> >         ../tools/syscalls/scripts/../core/syscall.lua:259: in function 'core.syscall.addArgs'
> >         ../tools/syscalls/scripts/../core/syscall.lua:362: in function 'core.syscall.add'
> >         ../tools/syscalls/scripts/../core/freebsd-syscall.lua:62: in function 'core.freebsd-syscall.parseSysfile'
> >         ../tools/syscalls/scripts/../core/freebsd-syscall.lua:134: in function 'core.freebsd-syscall.new'
> >         ../tools/syscalls/scripts/syscalls.lua:90: in main chunk
> >         [C]: in ?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Catching this error might be with doing.
> 
> I think it is because `)` is unescaped in the `gsub`:
> 
> ```lua
> > scarg = ")"
> > scarg:gsub(scarg:match("([^* ]+)$") .. "$", "")
> stdin:1: invalid pattern capture
> stack traceback:
> 	[C]: in function 'string.gsub'
> 	stdin:1: in main chunk
> 	[C]: in ?
> ```
> 
> One simple fix is to escape it:
> 
> ```lua
> > scarg:gsub(scarg:match("([^* ]+)$"):gsub("([^%w])", "%%%1") .. "$", "")
> 	1
> ```
> 
> **EDIT**: Sorry, the first workaround did not work. The current one should work (TM).

Yes, this was indeed the issue. Instead of escaping it though, I pcall'ed the gsub and print error if the pcall fails. If we're picking up a Lua escape sequence, then I figure it's safe to assume that's a malformed system call declaration.

Thanks!

I tested with that declaration and it's aborting with error: `Malformed argument line: )`

agge3 commented on sys/tools/syscalls/tools/generator.lua:
```diff
@@ -0,0 +1,117 @@
+	local comment_middle = comment or "*"
+	local comment_end = comment or "*/"
+
+	-- Don't enter loop if it's the simple case.
```
`generator:preamble()` has been cleaned up and made less unnecessarily complex, following this suggestion and others.

brooksdavis reviewed (commented):
I've updated my branch with a bunch of whitespace fixes and a few fixups of my prior commits. Except for the extra file generation bits the commits on my branch can be seen as reviews in commit form and squashed directly into your commit without credit. The other two add enough content that a Co-authored-by: credit might be nice if squashed.

https://github.com/brooksdavis/freebsd/tree/refactor-makesyscalls.lua

brooksdavis commented:
I've added a proposed rebasing and squashing of the new code with follow on commits to connect the new code at https://github.com/brooksdavis/freebsd/tree/refactor-makesyscalls.lua-rebased. Happy to take any feedback about the organization or contents.

If it looks good, I'll plan to push to FreeBSD once the current stabilization week ends.

agge3 commented on sys/tools/syscalls/config.lua:
```diff
@@ -0,0 +1,290 @@
+					config[k] = v
+				end
+				-- Construct config modified table as config is processed.
+				config.modifications[k] = true
```
Yes, the intended behavior was an else branch. I see that you've changed this. Thank you!

agge3 commented on sys/tools/syscalls/core/freebsd-syscall.lua:
```diff
@@ -0,0 +1,124 @@
+			if s:add(line) then
+				-- Append to system call list.
+				for t in s:iter() do
+					if t:validate(t.num - 1) then
```
This is a relic of things being more fragile on earlier iterations. You are correct in this assumption. `validate()` is now called in both locations.

agge3 commented:
> I've added a proposed rebasing and squashing of the new code with follow on commits to connect the new code at https://github.com/brooksdavis/freebsd/tree/refactor-makesyscalls.lua-rebased. Happy to take any feedback about the organization or contents.
> 
> If it looks good, I'll plan to push to FreeBSD once the current stabilization week ends.

Sorry I didn't follow-up more properly on your previous work. Last week, and this week, are very busy for me. I'll be less busy after this week.

This is great!!! Thank you for your help on the final stretch! I appreciate the work!! I'm glad you are pleased with the library. I'm happy to contribute and I really value the effort you've put into this.

I've merged [505d596](https://github.com/freebsd/freebsd-src/commit/505d596e80dfe0c64fa0709d15ea7bf614442f9f) from that branch to be part of this commit.

On my part, I've went through everything thoroughly. All changes are minor or non-invasive. Some final changes that I'd like to add (currently squashed into this commit):
 * Commit message includes Kyle as co-author. Him and Warner were both my mentors during GSoC, so that seems fitting with Warner as a co-author.
 * Removed `syscall:comment()` as it's broken/unimplemented and there's no longer a use for it.
 * Removed other `XXX` remarks that are implementation relics and should not be included.
 * Added `syscall:validate()` to both branches of `FreeBSDSyscall:parseSysfile()`. That's correct and numbers should be validated in either case.
 * `syscall_mk` has config as a parameter once again, the same as you've done with `libsys.h` and `syscalls.map`. Luacheck complains about unused parameter. Having config available in every module is more intuitive and semantically makes sense.
 * enforced `style.lua(9)` on outside credited functions, for consistency.
 * Punctuation (periods) were inconsistent in comments; comment punctuation is now consistent.
 * Semantics in comments are more consistent. I opted for Doxygen-style comments in some places, without a strict enforcement everywhere. I see that you continued the Doxygen-style. I've made comments more consistent, but this divide still exists.

Additionally, I noticed that "converstion" in DTrace generation is misspelled. I've changed to be spelled correctly. Considering how minor, now seems like a good time to change that. Pending feedback, I've left that as a second commit on this PR (which can be squashed).

On the weekend (before next week), I'm hoping to go through again. But I'm happy with this being my final input.

In terms of this PR, would you like me to pull your commits over, or would you prefer I leave it how it is?

brooksdavis commented:
I've rebased my branch with your changes (including the typo fix squashed into the main commit). I also fixed the typo upstream to reduce diffs in generated files.

I've added commits removing capabilities.conf and `capenabled=` support. I think it's ok to commit with the support and them immediately remove it on the grounds that someone might be using them downstream and this would make it easier to restore that functionality temporarily.

I'm happy to do another update Monday if you want to take another pass through the code over the weekend. It's probably somewhat easier if you don't incorporate my changes, but if there are any you want to take as part of the main commit, go ahead.

agge3 commented:
> I've rebased my branch with your changes (including the typo fix squashed into the main commit). I also fixed the typo upstream to reduce diffs in generated files.
> 
> 
> 
> I've added commits removing capabilities.conf and `capenabled=` support. I think it's ok to commit with the support and them immediately remove it on the grounds that someone might be using them downstream and this would make it easier to restore that functionality temporarily.
> 
> 
> 
> I'm happy to do another update Monday if you want to take another pass through the code over the weekend. It's probably somewhat easier if you don't incorporate my changes, but if there are any you want to take as part of the main commit, go ahead.

Great. Sounds good!

Yes, I'll take another pass over the weekend.

I'll report any changes I have (if any at all), so everything's ready for Monday.

brooksdavis commented:
Merged as commit 9ded074e875c29cb92d5f643801990d7bb23cca4
